File: lua/puzzle4-2.lua
require "utils"

local cards = {}
for line in io.lines() do
	local nums = string.split(line, ":")[2]
	assert(nums)
	local winning, ours = unpack(string.split(nums, "|"))
	assert(winning and ours)

	local win_nums = {}
	winning = string.split(string.strip(winning))
	for _, num in ipairs(winning) do
		win_nums[num] = 1
	end
	ours = string.split(string.strip(ours))

	local wins = 0
	for _, num in ipairs(ours) do
		if win_nums[num] then wins = wins + 1 end
	end
	table.insert(cards, {n=1, wins=wins})
end

local sum = 0
for i, card in ipairs(cards) do
	sum = sum + card.n
	for j = 1, card.wins do
		if cards[i + j] then
			cards[i + j].n = cards[i + j].n + card.n
		end
	end
end
print(sum)

File: rust/src/bin/puzzle4-2.rs
use std::collections::HashMap;

fn main() {
    let mut counts: HashMap<usize, u32> = HashMap::new();
    let mut last = 0;
    for (i, line) in std::io::stdin().lines().map(|s| s.unwrap()).enumerate() {
        *counts.entry(i).or_insert(0) += 1; // count original
        let cur_count = counts[&i];
        let (_, numbers) = line.split_once(':').unwrap();
        let (winning, ours) = numbers.split_once('|').unwrap();
        let winning: HashMap<&str, usize> = winning
            .split_ascii_whitespace()
            .zip([1].into_iter().cycle())
            .collect();
        let wins: usize = ours
            .split_ascii_whitespace()
            .filter_map(|n| winning.get(&n))
            .sum();
        for j in 1..=wins {
            *counts.entry(i + j).or_insert(0) += cur_count;
        }
        last = i;
    }
    let total = counts
        .iter()
        .map(|(&k, &v)| if k <= last { v } else { 0 })
        .sum::<u32>();
    println!("{total}");
}
